src/juju/watchers.ts

```diff
@@ -1,24 +1,58 @@
-import { AllWatcherDelta, ModelWatcherData } from "./types";
+import mergeWith from "lodash.mergewith";
+
+import type { AllWatcherDelta, ModelData, ModelWatcherData } from "./types";
+
+function generateModelWatcherBase(): ModelData {
+  return {
+    model: {
+      "cloud-tag": "",
+      "controller-uuid": "",
+      "is-controller": false,
+      "model-uuid": "",
+      config: {},
+      constraints: {},
+      life: "",
+      name: "",
+      owner: "",
+      region: "",
+      sla: {
+        level: "",
+        owner: "",
+      },
+      status: {
+        current: "",
+        message: "",
+        since: "",
+        version: "",
+      },
+      type: "",
+      version: "",
+    },
+  };
+}
 
 export function processDeltas(
-  state: ModelWatcherData,
+  modelWatcherData: ModelWatcherData,
   deltas: AllWatcherDelta[]
 ) {
   if (!deltas) {
     return;
   }
-  // console.log(deltas);
   deltas.forEach((delta) => {
+    // The delta messages are in the format [entity, action, data].
```

Comment:
You could destructure the array so you wouldn't need to refer to this comment if you were further down the file, say.
``` typescript
const [entity, action, data] = delta;
```

Reply:
Good call I'll do this in the follow-up as I've already moved on from this one :)

Reply:
In doing this as a follow-up I found what I believe to be a TypeScript 'bug'...

I added another case for the `charm` deltas as follows...

```ts
...
      case "charm":
        switch (delta[1]) {
          case "change":            
            const formatted = {
              [delta[2]["charm-url"]]: delta[2],
            };
            mergeWith(modelWatcherData[modelUUID].charms, formatted);
            break;
        }
        break;
...
```

now changing `delta[2]` to `data` as per `const [entityType, actionType, data] = delta`

```ts
...
            const formatted = {
              [delta[2]["charm-url"]]: delta[2],
            };
...
```

gives the TS error...

```
Element implicitly has an 'any' type because expression of type '"charm-url"' can't be used to index type 'ModelChangeDelta | ActionChangeDelta | ApplicationChangeDelta | CharmChangeDelta | UnitChangeDelta | MachineChangeDelta | RelationChangeDelta'.
  Property 'charm-url' does not exist on type 'ModelChangeDelta | ActionChangeDelta | ApplicationChangeDelta | CharmChangeDelta | UnitChangeDelta | MachineChangeDelta | RelationChangeDelta'.ts(7053)
```

So it appears that it cannot correctly infer the available data if it's been deconstructed as it only thinks it has a single key, `model-uuid`.